### PR TITLE
DAS-2276: retain 3 and 4 band information in browse images

### PR DESCRIPTION
**Pull request by flamingbear**
## Description

Code is modified to change the handling of 3 and 4 band input GeoTIFFs.

`convert_mulitband_to_raster` is modified so that 4 band images are only modified:
    1. If the original dtype is not 'uint8' (very rare) and the max valid value
       exceeds 255, then the data is scaled from 0 to 255
    2. If Rasterio picked up any missing values, they will be set to 0. This
       shouldn't happen because 4 bands would use the alpha layer, but also, if
       an rgb band has a nan, it's probably because the alpha is 0 (TRANSPARENT)

`convert_mulitband_to_raster` is modified so that 3-band images are modified:
    1. To add a transparent alpha layer for any missing data values.
    2. To scale if the input dtype is not 'uint8'

## Jira Issue ID

[DAS-2276](https://bugs.earthdata.nasa.gov/browse/DAS-2276)

## Local Test Steps

**Before you build or test this PR do these steps first!**

Pull current hybig docker image:
```
> docker pull ghcr.io/nasa/harmony-browse-image-generator:2.0.2
```

set Harmony-In-A-Box to use it by putting this in your harmony/.env file:
```
HYBIG_IMAGE=ghcr.io/nasa/harmony-browse-image-generator:2.0.2
HYBIG_SERVICE_QUEUE_URLS='["ghcr.io/nasa/harmony-browse-image-generator:2.0.2,http://sqs.us-west-2.localhost.localstack.cloud:4566/000000000000/hybig.fifo"]'
```
also enable opera-rtc-s1-browse by including it in your `LOCALLY_DEPLOYED_SERVICES=..[snip]....,net2cog,harmony-smap-l2-gridder,opera-rtc-s1-browse`

Pull the opera-rtc docker image (If you're on an M1 machine you'll need to force the `amd64` image.):
```
docker pull --platform linux/amd64 ghcr.io/asfhyp3/opera-rtc-s1-browse:latest
```

Get on the vpn(?) and download this file, you may need to accept the EULA.
https://datapool-test.asf.alaska.edu/RTC/OPERA-S1/OPERA_L2_RTC-S1_T035-073251-IW2_20240512T020817Z_20240512T122756Z_S1A_30_v1.0_VV.tif

Fire up your Harmony-In-A-Box... Ensure the pods for opera-rtc-s1-browse are running correctly, and run the DAS-2276-validate.py script

copy the `.png`, `.pgw` and `.png.aux.xml` into an original directory.

Verify that the .png is paletted with `gdalinfo`, you should see a palette and one band:
```
<snip>
Center      (-122.3788455,  37.8443403) (122d22'43.84"W, 37d50'39.62"N)
Band 1 Block=4282x1 Type=Byte, ColorInterp=Palette
  Color Table (RGB with 256 entries)
    0: 255,255,255,255
    1: 248,255,248,255
    2: 237,255,237,255
    3: 227,255,227,255
<snip>
```

**Now you can verify the PR**

Pull this branch, build and run tests:
```
❯ ./bin/build-image && ./bin/build-test &&  ./bin/run-test
```

Update your harmony .env so Harmony-In-A-Box will run the new locally-built version of hybig:
```
HYBIG_IMAGE=ghcr.io/nasa/harmony-browse-image-generator:latest
HYBIG_SERVICE_QUEUE_URLS='["ghcr.io/nasa/harmony-browse-image-generator:latest,http://sqs.us-west-2.localhost.localstack.cloud:4566/000000000000/hybig.fifo"]'
```

Kill and restart Harmony-In-A-Box to pick up the changes.

Re-run the validation script. Copy the files to a different directory. run `gdalinfo` again to verify that the images are rgba banded.

```
<snip>
Center      (-122.3788455,  37.8443403) (122d22'43.84"W, 37d50'39.62"N)
Band 1 Block=4282x1 Type=Byte, ColorInterp=Red
  Mask Flags: PER_DATASET ALPHA
Band 2 Block=4282x1 Type=Byte, ColorInterp=Green
  Mask Flags: PER_DATASET ALPHA
Band 3 Block=4282x1 Type=Byte, ColorInterp=Blue
  Mask Flags: PER_DATASET ALPHA
Band 4 Block=4282x1 Type=Byte, ColorInterp=Alpha
<snip>
```

## PR Acceptance Checklist
* [X] Jira ticket acceptance criteria met.
* [X] `CHANGELOG.md` updated to include high level summary of PR changes.
* [X] `docker/service_version.txt` updated if publishing a release.
* [X] Tests added/updated and passing.
* [X] Documentation updated (if needed).

### CODE REVIEW

File: hybig/sizes.py [hunk 1]
```diff
@@ -422,7 +422,7 @@ def find_closest_resolution(
 
     """
     best_info = None
-    smallest_diff = np.Infinity
+    smallest_diff = np.inf
```
Comment (flamingbear) [inline hybig/sizes.py:425]: This should have been in #38 😬

File: tests/test_service/test_adapter.py [hunk 2]
```diff
@@ -270,10 +270,14 @@ def move_tif(*args, **kwargs):
             mock_reproject.call_args_list, expected_reproject_calls
         ):
             np.testing.assert_array_equal(
-                actual_call.kwargs['source'], expected_call.kwargs['source']
+                actual_call.kwargs['source'],
+                expected_call.kwargs['source'],
+                strict=True,
```
Comment (flamingbear) [inline tests/test_service/test_adapter.py:275]: adding `strict=True` forces the comparison to "[raise an AssertionError when either the shape or the *data type* of the array_like objects does not match](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html)."

File: tests/unit/test_browse.py [hunk 3]
```diff
@@ -553,18 +604,38 @@ def test_convert_4_multiband_masked_to_raster(self):
         )
         ds.to_numpy.return_value = input_array
 
-        expected_raster = np.ma.array(
-            data=[
-                [[0, 0, 0, 121], [0, 0, 0, 64], [0, 0, 255, 0], [0, 0, 13, 255]],
-                [[0, 0, 0, 255], [0, 0, 0, 255], [0, 0, 255, 255], [0, 0, 255, 255]],
-                [[0, 0, 0, 121], [0, 0, 0, 64], [0, 0, 255, 0], [0, 0, 13, 255]],
-                [[0, 0, 0, 255], [0, 0, 0, 255], [0, 0, 255, 255], [0, 0, 255, 255]],
+        expected_raster = np.array(
```
Comment (flamingbear) [inline tests/unit/test_browse.py:607]: These values are no longer scaled.

**Review by owenlittlejohns — COMMENTED** [review]
I've left a couple of nits and questions. I'll try running the test instructions later this afternoon, though. Generally, this makes sense at first glance (beyond not quite knowing the initial need to make 3 or 4 band inputs not-paletted).

**Comment by flamingbear** [comment]
> need to make 3 or 4 band inputs not-paletted
@owenlittlejohns This was given to me as a GIBS requirement because of how they handle multiple images displayed at once.  Alexy explained it, but I cannot regurgitate it back to you. Other than it did make some sense in terms of the number of palettes they had to handle when displaying multiple images and how the quantization mechanism was generating a different palette for every image (as expected and designed)

File: CHANGELOG.md [hunk 4]
```diff
@@ -90,7 +91,8 @@ outlined by the NASA open-source guidelines.
 For more information on internal releases prior to NASA open-source approval,
 see legacy-CHANGELOG.md.
 
-[unreleased]:https://github.com/nasa/harmony-browse-image-generator/compare/2.0.2..HEAD
+[unreleased]:https://github.com/nasa/harmony-browse-image-generator/compare/2.1.0..HEAD
+[v2.1.0]:https://github.com/nasa/harmony-browse-image-generator/compare/2.0.2..2.1.0
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply CHANGELOG.md:95]: I did want to do that.

File: hybig/browse.py [hunk 5]
```diff
@@ -233,26 +237,49 @@ def convert_mulitband_to_raster(data_array: DataArray) -> ndarray:
...
-    raster = np.nan_to_num(np.around(norm(bands) * 255.0), copy=False, nan=0.0).astype(
-        'uint8'
-    )
+    return np.concatenate((raster, nan_alpha[None, ...]), axis=0)
+
+
+def convert_to_uint8(bands: ndarray, dtype: str | None) -> ndarray[uint8]:
+    """Convert Banded data with NaNs (missing) into a uint8 data cube.
+
+    99.99% of the time this will simply pass through the data coercing it back
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py:255]: Most didn't sound like often enough.   Almost always I guess I could say.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,10 +4,11 @@ HyBIG follows semantic versioning. All notable changes to this project will be
 documented in this file. The format is based on [Keep a
 Changelog](http://keepachangelog.com/en/1.0.0/).
 
-## [unreleased] - 2024-12-10
+## [v2.1.0] - 2024-12-13
 
 ### Changed
 
+* Input GeoTIFF RGB[A] images are **no longer palettized** when converted to a PNG. The new resulting output browse images are now 3 or 4 band PNG retaining the color information of the input image.[#39](https://github.com/nasa/harmony-browse-image-generator/pull/39)
```
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply CHANGELOG.md:11]: Okay - I think that's what I needed to hear: that GIBS was asking for this and that they had helped determine the best path forward. Thanks!

File: hybig/sizes.py (hunk 1 above)
Comment (flamingbear) [reply hybig/sizes.py:425]: Hmm.  I don't know now.  I thought I had an error because of it. but I can't even find one for lint or ruff now. I think it's deprecated in any case. https://numpy.org/devdocs/numpy_2_0_migration_guide.html

File: hybig/browse.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py]: 8d8968c

File: hybig/browse.py [hunk 7]
```diff
@@ -331,15 +358,19 @@ def image_driver(mime: str) -> str:
 
 
 def prepare_raster_for_writing(
-    raster: ndarray, driver: str
+    raster: ndarray,
+    driver: str,
+    input_bands: int,
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py:363]: No I refactored this a couple of times but didn't update the comment or what that var was.  You have another comment in here and I'll say more there.

File: hybig/browse.py [hunk 8]
```diff
@@ -331,15 +358,19 @@ def image_driver(mime: str) -> str:
...
 ) -> tuple[ndarray, dict | None]:
     """Remove alpha layer if writing a jpeg."""
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py]: Yes I agree, it's not actually more complicated it's not really named well maybe?

I waffled a couple of times because this logic could just be a few lines at L184, but that felt like it cluttered up the flow there.

This is just kind of hiding a few different things.

It's dropping an alpha layer if we're writing a JPEG
It's palettizing a single-band input PNG
It's setting the color_palette to None 

I can probably update the comment to spell that out.

[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py]: Ok so I did think through this abit. 41bad90 is the commit that updates the comment. and b6aadd0 renames the function.  I don't love either of those completely.

I think that since now we're only palettizing single bands, that convert_singleband_to_raster and then the palettizing after is not super useful and I'm going to be looking at that next. I expect this function to change or go away after that.

can we leave this as updated in these two commits for now?

File: hybig/browse.py [hunk 9]
```diff
@@ -331,15 +358,19 @@ def image_driver(mime: str) -> str:
...
+
+    if driver == 'PNG' and input_bands == 1:
+        # we only paletize single band input data
+        return palettize_raster(raster)
 
-    return palettize_raster(raster)
+    return raster, None
 
 
 def palettize_raster(raster: ndarray) -> tuple[ndarray, dict]:
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py:393]: I just left a comment saying I'm not sure that's valuable anymore now that we're leaving 3/4 bands as is.  

That said, this is actually an accurate comment for what this function is doing. `"""Convert an RGB or RGBA image into a 1band image and palette.` 

You are confused by the comment on L387 which I will update to mention it's a 4band converted raster.

[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py:393]: bae52d6. 

Cleaning this definitely makes me think the normalization of single band input to rasters so we treat everything the same has outgrown it's usefulness. But I will not address in this PR

File: hybig/browse.py [hunk 10]
```diff
@@ -331,15 +358,19 @@ def image_driver(mime: str) -> str:
...
+    driver: str,
+    input_bands: int,
 ) -> tuple[ndarray, dict | None]:
     """Remove alpha layer if writing a jpeg."""
-    if driver == 'JPEG':
-        if raster.shape[0] == 4:
-            raster = raster[0:3, :, :]
-        return raster, None
+    if driver == 'JPEG' and raster.shape[0] == 4:
+        return raster[0:3, :, :], None
```
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply hybig/browse.py:384]: Discussed in Slack. Unimportant, and fine with no changes being made.

File: hybig/browse.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply hybig/browse.py:394]: Okay, I think the inline comment you added saying that the single band has been converted probably covers my confusion. Thanks for the clarification.

File: hybig/browse.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply hybig/browse.py]: I think this looks clearer now. I think the function name and `band_count` are both clearer as to what the function and variable are for.

I think the key thing in the description of the colour map return value is that you mention "palette indices", which helps me know that I should only expect the colour map for a paletted PNG. (This all probably seems obvious as I type it out, but I'm thinking of when I review a different PR in six months and look at this function again)

TL;DR - I think this does the job and is clearer than before.

**Review by owenlittlejohns — COMMENTED** [review]
The code changes look good. I was unable to verify that the updated service runs as expected due to issues with the source data used in the test (this is being discussed internally in Slack).

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (flamingbear) [reply CHANGELOG.md:11]: Nope.  Disagree, but I swap them pretty intermittently. [Palette](https://en.wikipedia.org/wiki/Palette_(computing)) is the set of available colors. So the completely made up word is palettized when taking an RGBA image to value and color table lookup.  (and it still looks wrong)

[the remark this answers is not in the document]
Comment (lyonthefrog) [reply CHANGELOG.md:11]: Alright! I accept this.

File: hybig/browse.py [hunk 11]
```diff
@@ -330,16 +357,38 @@ def image_driver(mime: str) -> str:
...
+            - prepared_raster is the processed ndarray
+            - color_map is either None or a dict mapping palette indices to RGBA values
+
+
+    """
+    if driver == 'JPEG' and raster.shape[0] == 4:
+        return raster[0:3, :, :], None
+
+    if driver == 'PNG' and band_count == 1:
+        # Only paletize single band input data that has been converted to an
```
[the remark this answers is not in the document]
Comment (lyonthefrog) [reply hybig/browse.py:387]: Honestly no idea about this.

File: hybig/browse.py [hunk 12]
```diff
@@ -233,26 +237,49 @@ def convert_mulitband_to_raster(data_array: DataArray) -> ndarray:
...
+
+
+def original_dtype(data_array: DataArray) -> str | None:
+    """Return the original input data's type.
+
+    rastero_optn retains the input dtype in the encoding dictionary and is used
+    to understand what kind of casts are safe.
+
+    """
+    return data_array.encoding.get('dtype') or data_array.encoding.get('rasterio_dtype')
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py:282]: It's a [boolean or](https://docs.python.org/3/reference/expressions.html#boolean-operations) "The expression x or y first evaluates x; if x is true, its value is returned; otherwise, y is evaluated and the resulting value is returned."

File: hybig/browse.py [hunk 13]
```diff
@@ -233,26 +237,49 @@ def convert_mulitband_to_raster(data_array: DataArray) -> ndarray:
...
+        raster = np.nan_to_num(bands).astype('uint8')
 
-    return np.concatenate((raster, alpha[None, ...]), axis=0)
+    return raster
+
+
+def original_dtype(data_array: DataArray) -> str | None:
+    """Return the original input data's type.
+
+    rastero_optn retains the input dtype in the encoding dictionary and is used
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py:278]: It's sort of used directly, because data_array is the return value of `rasterio_open()`
This is just a convenience function to know what the datatype was on the original input file.

File: tests/unit/test_browse.py (hunk 3 above)
Comment (flamingbear) [reply tests/unit/test_browse.py:608]: Well, yes and no.  Yes, they are all in the valid range so they aren't scaled, but they're no longer scaled, because the code before this change did scale all 4 band rasters https://github.com/nasa/harmony-browse-image-generator/blob/2.0.2/hybig/browse.py#L237-L243

File: tests/unit/test_browse.py [hunk 14]
```diff
@@ -581,33 +653,59 @@ def test_convert_5_multiband_to_raster(self):
...
+
+        expected, _ = standardize_raster_for_writing(raster, driver, count)
+        np.testing.assert_array_equal(raster, expected, strict=True)
+        palettize_mock.assert_not_called()
+
+    @patch('hybig.browse.palettize_raster')
+    def test_standardize_raster_for_writing_png_3band(self, palettize_mock):
+        raster = self.random.integers(255, size=(3, 7, 8))
+        driver = 'PNG'
+        count = 'not 1'
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply tests/unit/test_browse.py:695]: exactly.  I could have put 0 or 4 but this is explicitly saying not 1

File: hybig/browse.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (flamingbear) [reply hybig/browse.py]: 6ad622f

File: tests/unit/test_browse.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (lyonthefrog) [reply tests/unit/test_browse.py:695]: I actually like this, it's much easier to read the test

**Review by lyonthefrog — APPROVED** [review]

**Review by owenlittlejohns — APPROVED** [review]
